From 58bda4840a1e3b446feed5d722370632a9896819 Mon Sep 17 00:00:00 2001
From: Stephen Smalley <stephen.smalley.work@gmail.com>
Date: Wed, 2 Oct 2024 16:52:15 -0400
Subject: [PATCH 48/90] selinux: limit selinux netlink notifications to init
 namespace

To prevent modifying the enforcing status or resetting the AVC
in the userspace policy enforcers in another SELinux namespace,
do not deliver SELinux netlink notifications unless they were
generated by the init SELinux namespace. Child SELinux namespaces
are expected to always use the SELinux status page instead,
which is the default with modern libselinux on modern kernels.

Signed-off-by: Stephen Smalley <stephen.smalley.work@gmail.com>

Upstream-commit: 57f9b6f0a0f0bb71062a908b3b54b6fbc6d877ea
---
 security/selinux/hooks.c            |  2 +-
 security/selinux/include/security.h |  1 +
 security/selinux/netlink.c          | 10 ++++++++++
 3 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index 2f4a9c416d85a2c391f83d347d90237b7fa2931c..d4d4a568da5cc82dc3ee2758df6173ceaafbc52d 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -107,7 +107,7 @@
 
 #define SELINUX_INODE_INIT_XATTRS 1
 
-static struct selinux_state init_selinux_state;
+struct selinux_state init_selinux_state;
 struct selinux_state *current_selinux_state;
 
 /* SECMARK reference count */
diff --git a/security/selinux/include/security.h b/security/selinux/include/security.h
index 01f2476bddf6bfc08809b0d0164d9d26be99470f..6b658203c7a8b3874ae8f95af3d6400eed775b1d 100644
--- a/security/selinux/include/security.h
+++ b/security/selinux/include/security.h
@@ -109,6 +109,7 @@ struct selinux_state {
 
 void selinux_avc_init(struct selinux_avc **avc);
 
+extern struct selinux_state init_selinux_state;
 extern struct selinux_state *current_selinux_state;
 
 static inline bool selinux_initialized(const struct selinux_state *state)
diff --git a/security/selinux/netlink.c b/security/selinux/netlink.c
index 1760aee712fd2ef4ed805d6f0a9edffedd547e58..4b8983f2b9f392044f5fa6f51d1405a161ed296e 100644
--- a/security/selinux/netlink.c
+++ b/security/selinux/netlink.c
@@ -71,6 +71,16 @@ static void selnl_notify(int msgtype, void *data)
 	struct sk_buff *skb;
 	struct nlmsghdr *nlh;
 
+	/*
+	 * Suppress SELinux netlink notifications from a non-init selinux
+	 * namespace. Otherwise, a child namespace could confuse an ancestor
+	 * namespace with its own enforcing mode or policy reload
+	 * notifications. Modern libselinux on modern kernels uses the
+	 * SELinux status page instead, which is per-namespace already.
+	 */
+	if (current_selinux_state != &init_selinux_state)
+		return;
+
 	len = selnl_msglen(msgtype);
 
 	skb = nlmsg_new(len, GFP_USER);
